Refactor Card and FactCheckCard

dariusk opened the pull request:
I'm considering `Card` now to be an empty rectangle container that can be passed a `cardUrl` and a `footer`, and highlights when you mouse over it. The card manages an `isHovered` state that is provided to its `children` via `Context` as needed. So in this case `FactCheckCard` renders a `Card` with some children in the correct layout. The `ItemDescription` is wrapped with the `CardHoverContext.Consumer`, because `ItemDescription` needs to know that `Card` (not `FactCheckCard`!) is being hovered. In other cases I would do this by passing state from Card to its children, but since we have a generic `children` object I'm doing it this way instead. Seems to work well.

References: CV2-3991

## Type of change

- [X] Performance improvement and/or refactoring (non-breaking change that keeps existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Security mitigation or enhancement
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Automated test (add or update automated tests)

caiosba reviewed (approved):
Code looks good! I like that now `<Card />` is way simpler. I tested it locally with the fact-checks cards and it works well (I had to push a fix though, which I talked to @dariusk before). The only issue I noticed is that ellipsis are now missing for long URLs:

![Captura de tela de 2023-12-22 17-54-47](https://github.com/meedan/check-web/assets/117518/68f16f74-d268-4119-a89d-6ece9ee92f44)

But I'm approving for now in order to unblock, but please fix it in the next PR, @dariusk.

dariusk commented:
@caiosba I made a bunch of updates -- please see CV2-3991 comments for details but basically I have added:

 * sorted list of platforms that requests came from for a `SharedItemCard` (called `ItemChannels`)
 * helper function for compact number formatting (and longer number formatting)
 * conditional media icon when the media count is 1
 * tooltips for workspaces, dates, and counts - these are pretty complex and I explain more in CV2-3991

caiosba commented on src/app/components/cds/media-cards/ItemChannels.js:
```diff
@@ -0,0 +1,104 @@
+}) => {
+  const tiplines = CheckChannels.TIPLINE;
+
+  const humanTiplineNames = {
+    WHATSAPP: 'WhatsApp',
+    MESSENGER: 'Messenger',
+    TWITTER: 'X (Twitter)',
+    TELEGRAM: 'Telegram',
+    VIBER: 'Viber',
+    LINE: 'Line',
+    INSTAGRAM: 'Instagram',
+  };
+
+  function getTiplineNameFromChannelNumber(channel) {
+    return Object.entries(tiplines).find(item => item[1] === channel.toString())[0];
+  }
+
+  function getHumanNameFromChannelNumber(channel) {
+    const tipline = Object.entries(tiplines).find(item => item[1] === channel.toString())[0];
+    return humanTiplineNames[tipline];
+  }
+
+  function ChannelIcon({ channel }) {
+    const tipline = getTiplineNameFromChannelNumber(channel);
+
+    function tiplineIcon(name) {
+      switch (name) {
+      case 'WHATSAPP': return <WhatsAppIcon style={{ color: 'var(--whatsappGreen)' }} />;
+      case 'MESSENGER': return <FacebookIcon style={{ color: 'var(--facebookBlue)' }} />;
+      case 'TWITTER': return <TwitterIcon style={{ color: 'var(--xBlack)' }} />;
+      case 'TELEGRAM': return <TelegramIcon style={{ color: 'var(--telegramBlue)' }} />;
+      case 'VIBER': return <ViberIcon style={{ color: 'var(--viberPurple)' }} />;
+      case 'LINE': return <LineIcon style={{ color: 'var(--lineGreen)' }} />;
+      case 'INSTAGRAM': return <InstagramIcon style={{ color: 'var(--instagramPink)' }} />;
+      default: return null;
+      }
+    }
```
Question/suggestion: What if we move this mapping to https://github.com/meedan/check-web/blob/develop/src/app/CheckChannels.js? We do something similar in https://github.com/meedan/check-web/blob/develop/src/app/CheckFeedLicenses.js.

It can be out of scope for this ticket too - maybe we can just create a small refactoring ticket to abstract that to https://github.com/meedan/check-web/blob/develop/src/app/CheckChannels.js and replace it in other places it's done (like the channels filter).

dariusk replied:
Good idea. Filed CV2-4181 for the refactor.

caiosba commented on src/app/components/cds/media-cards/ItemChannels.js:
```diff
@@ -0,0 +1,104 @@
+ItemChannels.propTypes = {
+  className: PropTypes.string,
+  sortMainFirst: PropTypes.bool,
+  channels: PropTypes.exact({
```
I like `exact` instead of `shapeOf`!

caiosba commented on src/app/components/search/SearchResultsCards/SharedItemCard.js:
```diff
@@ -0,0 +1,193 @@
+  channels: PropTypes.exact({
+    main: PropTypes.number,
+    others: PropTypes.arrayOf(PropTypes.number),
+  }),
```
Missing `propTypes`:

- `mediaThumbnail`
- `mediaType`
- `workspaces`

caiosba reviewed (changes requested):
@dariusk , looks good! I added a couple of comments, and locally I noticed that a truncation bug (I think it's related to the URL) still happens (the URL should truncate with ellipsis):

![pr-1785-bug-1](https://github.com/meedan/check-web/assets/117518/23b4114a-407a-4f36-b6e5-7d9589aafa65)

dariusk commented:
> @dariusk , looks good! I added a couple of comments, and locally I noticed that a truncation bug (I think it's related to the URL) still happens (the URL should truncate with ellipsis):
> 
> ![pr-1785-bug-1](https://private-user-images.githubusercontent.com/117518/294953954-23b4114a-407a-4f36-b6e5-7d9589aafa65.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ4NDMyMjQsIm5iZiI6MTcwNDg0MjkyNCwicGF0aCI6Ii8xMTc1MTgvMjk0OTUzOTU0LTIzYjQxMTRhLTQwN2EtNGYzNi1iNmU1LTdkOTU4OWFhZmE2NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwOVQyMzI4NDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNGE5YzIyMzFlYWI4YzhjNzUwMzQ0MmMxMzdkMjg5MTMwMjhjNzRmY2E0NWNhZjJlYWNmMDM3MTZmNTk4ZDU1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.BhZz143b77JrLJC47Yj5BjStBRcp1-ita4dXS0K7uIA)

@caiosba It looks like this is on some part of the Check app -- this component is not anywhere except for `Sandbox` right now, and it appears to be truncating correctly there:

![image](https://github.com/meedan/check-web/assets/266454/59c95080-e378-4d20-8c19-d8b3431f54d8)

You'll notice in the above screenshot that when the width is low, 1) the ellipsis truncation happens, and 2) the elements stack in a responsive flex flow.

caiosba commented:
Thanks @dariusk , can you please try with a long URL? I see it working with long titles and summaries in the fact-check card too.

dariusk commented:
@caiosba Long url does break it, looking into it.

dariusk commented:
I fixed the bug with displaying fact check urls that are very long, and I merged @Caio Almeida's work for `CheckFeedDataPoints`, adding the logic for the different display states for `SharedFeedItemCard` based on the data composition of a shared feed:

![image](https://github.com/meedan/check-web/assets/266454/5dba88f1-c223-479b-bd7b-71303fe1e0e1)

I added three switches to the sandbox - two of them to toggle what kind of data points are in a shared feed, and one to toggle the number of fact checks between 1 and 12345.

caiosba commented on src/app/components/Sandbox.js:
```diff
@@ -486,6 +504,42 @@ const SandboxComponent = ({ admin }) => {
+                  checked={listItemFactCheckCount > 1}
+                  onChange={() => {
+                    if (listItemFactCheckCount === 1) {
+                      // eslint-disable-next-line
+                      console.log('~~~SET TO 12345');
```
Just a reminder to delete this before merging

caiosba commented on src/app/components/Sandbox.js:
```diff
@@ -486,6 +504,42 @@ const SandboxComponent = ({ admin }) => {
+                      console.log('~~~SET TO 12345');
+                      setListItemFactCheckCount(12345);
+                    } else {
+                      // eslint-disable-next-line
+                      console.log('~~~SET TO 1');
```
Just a reminder to delete this before merging

caiosba reviewed (approved):
Thanks @dariusk , looks good now!
